dali/operators/image/color/color_space_conversion.cu

```diff
@@ -115,109 +47,18 @@ void ColorSpaceConversion<GPUBackend>::RunImpl(DeviceWorkspace &ws) {
-      const ImageTypePair kGRAY_TO_BGR { DALI_GRAY, DALI_BGR };
-      const ImageTypePair kGRAY_TO_YCbCr { DALI_GRAY, DALI_YCbCr };
-
-      if (conversion == kRGB_TO_BGR || conversion == kBGR_TO_RGB) {
-        // RGB -> BGR
-        // BGR -> RGB
-        detail::ConvertRGBToBGR8uKernel<<<grid, block, 0, stream>>>(
-          input_data, output_data, total_size);
-      } else if (conversion == kRGB_TO_YCbCr) {
-        // RGB -> YCbCr
-        DALI_CHECK_NPP(
-          nppiRGBToYCbCr_8u_C3R(
-            input_data, nStepInput, output_data, nStepOutput, size));
-      } else if (conversion == kBGR_TO_YCbCr) {
-        // BGR -> YCbCr
-        // First from BGR to RGB
-        detail::ConvertBGRToRGB8uKernel<<<grid, block, 0, stream>>>(
-          input_data, output_data, total_size);
-        // Then from RGB to YCbCr
-        DALI_CHECK_NPP(
-          nppiRGBToYCbCr_8u_C3R(
-            output_data, nStepOutput, output_data, nStepOutput, size));
-      } else if (conversion == kRGB_TO_GRAY) {
-        // RGB -> GRAY
-        DALI_CHECK_NPP(
-          nppiRGBToGray_8u_C3C1R(
-            input_data, nStepInput, output_data, nStepOutput, size));
-      } else if (conversion == kBGR_TO_GRAY) {
-        // BGR -> GRAY
-        const Npp32f aCoefs[3] = {0.114f, 0.587f, 0.299f};
-        DALI_CHECK_NPP(
-          nppiColorToGray_8u_C3C1R(
-            input_data, nStepInput, output_data, nStepOutput, size, aCoefs));
-      } else if (conversion == kYCbCr_TO_BGR) {
-        // First from YCbCr to RGB
-        DALI_CHECK_NPP(
-          nppiYCbCrToRGB_8u_C3R(
-            input_data, nStepInput, output_data, nStepOutput, size));
-        // Then from RGB to BGR
-        detail::ConvertRGBToBGR8uKernel<<<grid, block, 0, stream>>>(
-          output_data, output_data, total_size);
-      } else if (conversion == kYCbCr_TO_RGB) {
-        // First from YCbCr to RGB
-        DALI_CHECK_NPP(
-          nppiYCbCrToRGB_8u_C3R(
-            input_data, nStepInput, output_data, nStepOutput, size));
-      } else if (conversion == kGRAY_TO_BGR || conversion == kGRAY_TO_RGB) {
-        // GRAY -> RGB / BGR
-        detail::ConvertGrayToRGB8uKernel<<<grid, block, 0, stream>>>(
-          input_data, output_data, total_size);
-      } else if (conversion == kGRAY_TO_YCbCr) {
-        // GRAY -> YCbCr
-        detail::ConvertGrayToYCbCr8uKernel<<<grid, block, 0, stream>>>(
-          input_data, output_data, total_size);
-      } else if (conversion == kYCbCr_TO_GRAY) {
-        // YCbCr -> GRAY
-        detail::ConvertYCbCrToGray8uKernel<<<grid, block, 0, stream>>>(
-          input_data, output_data, total_size);
-      } else {
-        DALI_FAIL("conversion not supported");
-      }
-    }
+  DALI_ENFORCE(layout == "HWC" || (layout.empty() && output_shape.sample_dim() == 3),
+               make_string("Unexpected layout: ", layout, " shape: ", output_shape,
+                           ". Expected data in HWC layout."));
```

Comment:
Can't we have a video or a volume? We're flattening other dimensions anyway. We're only interested in the the channel being the last one.

Reply:
Done